[ENH] Extended parameter estimator base class with y in `fit`

satvshr opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->

#### Reference Issues/PRs
Fixes #7694 
Dependent on #7766 

#### What does this implement/fix? Explain your changes.
Adds y to the base class for parameter estimators and changes all the param estimators to accommodate the changes to their parent class.

fkiraly reviewed (changes requested):
Great!

I think we do not need to add the two new tags - I would add instead `requires_y` with the default `False`, and assume all estimators accept `y`.

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -4,8 +4,8 @@
 Scitype defining methods:
-    fitting                - fit(X)
-    updating               - update(X)
+    fitting                - fit(X, y)
```
should be clear that this is optional

satvshr replied:
Updated

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -4,8 +4,8 @@
-    fitting                - fit(X)
-    updating               - update(X)
+    fitting                - fit(X, y)
+    updating               - update(y, X=None)
```
X is mandatory, y is optional

satvshr replied:
Made the changes in the docstring, and the function is doing exactly what you mentioned above.

satvshr commented:
> why are you removing the treatment of VectorizedDF?

Isn't it meant for forecasters rather than estimators? I thought it is unlikely to be passed to estimators so I removed it, could add it back if you'd like me to.

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -46,19 +45,35 @@ def _coerce_to_list(obj):
         "object_type": "param_est",  # type of object
-        "X_inner_mtype": "pd.DataFrame",  # which types do _fit/_predict, support for X?
+        "X_inner_mtype": "pd.DataFrame",  # which types do _fit/_predict support for X?
+        "y_inner_mtype": "pd.Series",  # which types do _fit/_predict support for y?
```
can we also have `pd.DataFrame` as the default? Or it will be confusing for implementers.

satvshr replied:
WIll add it

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -46,19 +45,35 @@ def _coerce_to_list(obj):
+        "X_inner_mtype": "pd.DataFrame",  # which types do _fit/_predict support for X?
+        "y_inner_mtype": "pd.Series",  # which types do _fit/_predict support for y?
         "scitype:X": "Series",  # which X scitypes are supported natively?
+        "scitype:y": "univariate",  # which y are fine: univariate/multivariate/both
```
this should be the same as `scitype:X`

satvshr replied:
A lot of the code I wrote was inspired by the entire codebase, and everywhere I looked scitype:y was either univariate, multivariate, or both

fkiraly replied:
yes

fkiraly replied:
the fields need to make sense as part of the design. In this case, consistency with `X` is important.

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -155,20 +171,17 @@ def fit(self, X):
-            For further details:
-                on usage, see forecasting tutorial examples/01_forecasting.ipynb
-                on specification of formats, examples/AA_datatypes_and_datasets.ipynb
+        X : sktime-compatible container
```
can you leave the references on the permissible formats?

This may have to be updated though with newer API references, see the docstring for `BaseForecaster.fit`

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -180,196 +193,290 @@ def fit(self, X):
         # checks and conversions complete, pass to inner fit
         #####################################################
-        self._fit(X=X_inner)
+        self._fit(X=X_inner, y=y_inner)
```
I think we need to be careful here!

Imagine current estimators in a 3rd party package where `_fit` has only `X` as argument. All those estimators would now break, because you are passing an argument that is not available!

We need to ensure that we do not pass `y` when it is not available in the inner method as an argument. I will refactor some code that already does this and link the PR here so you can depend on it.

We should also test somewhere that inner estimators without `y` are still supported.

One way is to just not change all the estimators that do not use `y`.

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -180,196 +193,290 @@ def fit(self, X):
-
-        # update internal X with the new X
-        self._update_X(X_inner)
+        if X is not None:
```
I do not think we need to update `y`, it is not mandatory.

satvshr replied:
so remove it completely?

fkiraly reviewed (changes requested):
Great start!

Two major change requests:

* I would suggest we keep the extension contract stable by making the `y` in `_fit` optional.
* there seem to be major changes in the `_check` and the `_update` function - I would expect that at least the part for `X` remains unchanged, which does not seem to be the case. Could you explain what you are doing? Or did you just copy the code from somewhere else? Which would not be a good idea.

satvshr commented on sktime/param_est/base.py:
```diff
@@ -180,196 +193,290 @@ def fit(self, X):
         self.check_is_fitted()
 
-        if X is None or (hasattr(X, "__len__") and len(X) == 0):
+        if y is None or (hasattr(y, "__len__") and len(y) == 0):
```
line 238 gives a warning that y was passed to update so I thought X over here is used to represent y.

satvshr commented:
> * there seem to be major changes in the `_check` and the `_update` function - I would expect that at least the part for `X` remains unchanged, which does not seem to be the case. Could you explain what you are doing? Or did you just copy the code from somewhere else? Which would not be a good idea.

Did not copy the code from anywhere, took inspiration from similar code throughout the codebase. `_update` as I Mentioned in the reviews above, has a comment that it was meant for y and not X so I changed that given we could supply y now, I thought we were updating using y and not X (as default) which was an oversight on my part.

For '_check' the only thing I changed for X was the `VectorizedDF` part which as I mentioned above seemed unnecessary for an estimators check method.

fkiraly reviewed (changes requested):
Thanks.

I think this is an improvement and addresses the optional `y` in the `_fit` call.

Some remaining change requests:

* since the use of `y` in plugins is now optional again, an you remove the added, unused `y`-s from `_fit`-s across the code base?
* you seem to be rewriting the docstrings for various methods, deleting explanations. Why are you doing this? This is making the docstrings worse as you remove useful explanation. Are you using an AI?
* it is not a good idea to change the base tag `X_inner_mtypes`. This will break many existing estimators in the case `pd.Series`. Please read about the tags you are changing in the API reference for a more detailed explanation.
    * there are some odd statements - what does it mean that "X is not square", for example?
* in this PR, can you only add functionality for `y`, please? Any other new features, such as checking for missing values - useful, but separate PR please.
* the checks for `X` and `y` are almost the same. It is not good practice to copy-paste code twice. Instead, write a checker that can be applied to `X` and `y`, both, and call it twice.

satvshr commented:
> * you seem to be rewriting the docstrings for various methods, deleting explanations. Why are you doing this? This is making the docstrings worse as you remove useful explanation. Are you using an AI?

Yes the docstrings were all AI-generated, I reverted most of them back to what they originally were.

> * it is not a good idea to change the base tag `X_inner_mtypes`. This will break many existing estimators in the case `pd.Series`. Please read about the tags you are changing in the API reference for a more detailed explanation.

The default itself was 'pd.DataFrame', I changed nothing there. (line 58 in the old file and 73 in the updated one.

>   * there are some odd statements - what does it mean that "X is not square", for example?

What I learnt was that pairwise estimators are meant to be square in shape, am I wrong there?

> * in this PR, can you only add functionality for `y`, please? Any other new features, such as checking for missing values - useful, but separate PR please.

Removed that from the new one.

satvshr commented:
@fkiraly checks seem to be failing because of the test I wrote, I think you merged it after merging another PR which did not remove the estimator from the EXCLUDED_TESTS_BY_TEST (#7622)

satvshr commented:
@fkiraly any idea why I may be getting this error?
```
FAILED sktime/tests/test_all_estimators.py::TestAllObjects::test_get_test_params_coverage[ParamFitterPipeline] - AssertionError: ParamFitterPipeline.get_test_params should return at least two test parameter sets, but only 1 found.
assert 1 > 1
 +  where 1 = len([{'param_est': FixedParams(param_dict={3: 42, 'a': 2}), 'transformers': [('foo', ExponentTransformer(power=2)), ('bar', ExponentTransformer())]}])
```

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -275,81 +352,104 @@ def _check_X(self, X=None):
 
-        X_inner_mtype = _coerce_to_list(self.get_tag("X_inner_mtype"))
-        # X_inner_scitype = mtype_to_scitype(X_inner_mtype, return_unique=True)
+        if y is not None and self.get_tag("requires_y") is False:
```
remove this exception

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -275,81 +352,104 @@ def _check_X(self, X=None):
+            FORBIDDEN_MTYPES = ["numpyflat", "pd-wide"]
+            X_inner = self._validate_data(ALLOWED_SCITYPES, FORBIDDEN_MTYPES, X, "X")
+
+        # If both X and y are provided, check that their lengths match.
```
remove this exception

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -406,12 +520,13 @@ def _update(self, X):
         -------
         self : reference to self
         """
-        # default to re-fitting if update is not implemented
```
restore this line

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -359,17 +459,29 @@ def _update_X(self, X):
+                self._y = update_data(self._y, y)
+
         if X is not None:
-            # unwrap X if VectorizedDF
```
restore these lines

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -240,25 +302,40 @@ def update(self, X):
-        # checks and conversions complete, pass to inner update
-        self._update(X=X_inner)
+        # Pass the checked and converted data to the estimator-specific _update logic
+        _safe_call(self._fit, args=(), kwargs={"X": X_inner, "y": y_inner})
```
fix this

fkiraly commented on sktime/param_est/base.py:
```diff
@@ -240,25 +302,40 @@ def update(self, X):
-
-        # update internal X with the new X
-        self._update_X(X_inner)
+        if y is not None:
```
unnecessary case distinction, deal with `y` in the inner function

fkiraly reviewed (changes requested):
@satvshr, thanks for the changes, I will finalize the PR.

fkiraly reviewed (approved):
Finalized now - there was an issue with the tests which masked some bugs: https://github.com/sktime/sktime/pull/7737
